### PR TITLE
Issues 653 635

### PR DESCRIPTION
**Pull request by tom-englert**
Fixing issues #653 and #635

### CODE REVIEW

File: PropertyChanged.Fody/PropertyWeaver.cs [hunk 1]
```diff
@@ -72,6 +72,10 @@ void InjectAtIndex(int index)
             }
         }
 
+        var setMethod = propertyData.PropertyDefinition.SetMethod;
+        if (setMethod.GetNonAbstractBaseMethod(out _))
+            return;
```
Comment (ltrzesniewski) [inline PropertyChanged.Fody/PropertyWeaver.cs:77]: You need to check if the base method is actually called. If it's not, you still need to insert the event invoker.

I added the following:
 - `public virtual int Property4 { get; set; }` in `BaseClass`
 - `public override int Property4 { get; set; }` in `DerivedClass`

And the following test fails:

```C#
[Fact]
public void DerivedPropertyNotCallingBasePropertyShouldRaiseEvents()
{
    var instance = testResult.GetInstance("DerivedClass");
    instance.Property4 = 42;

    var expected = new[] { "Property4" };

    Assert.Equal(expected, (IList<string>)instance.ChangedEvents);
}
```

**Review by ltrzesniewski — COMMENTED** [review]
I like the approach :+1:

But I found a breaking issue, see the comment below. I suppose more tests would be nice.

File: PropertyChanged.Fody/PropertyWeaver.cs (hunk 1 above)
Comment (tom-englert) [reply PropertyChanged.Fody/PropertyWeaver.cs:77]: I would not consider this an error. 
If you override a property without calling the base classes methods, you intend to overrude the original behavior with something different - so here you give up property change notification.

Comment (tom-englert) [reply PropertyChanged.Fody/PropertyWeaver.cs:77]: And yes, more tests should be added. Maybe @DaleStan can jump in and port his tests, if we agree on this approach.

Comment (tom-englert) [reply PropertyChanged.Fody/PropertyWeaver.cs]: @ltrzesniewski now handling this use case, and added more tests.

File: PropertyChanged.Fody/CecilExtensions.cs [hunk 2]
```diff
@@ -11,6 +11,26 @@ public static string GetName(this PropertyDefinition propertyDefinition)
...
+        {
+            return false;
+        }
+
+        if (!(instruction.Operand is MethodReference methodReference))
+        {
+            return false;
+        }
+
+        if (methodReference.Resolve() != method)
```
Comment (ltrzesniewski) [inline PropertyChanged.Fody/CecilExtensions.cs:26]: Adding `if (methodReference.Name != method.Name) return false;` before the call to `Resolve` could be a nice optimization.

File: PropertyChanged.Fody/IlGeneratedByDependencyReader.cs [hunk 3]
```diff
@@ -16,16 +16,10 @@ public IlGeneratedByDependencyReader(TypeNode node)
...
-        {
-            methodComparer = GenericMethodComparer;
-            fieldComparer = GenericFieldComparer;
-        }
-        else
-        {
-            methodComparer = NonGenericMethodComparer;
-            fieldComparer = NonGenericFieldComparer;
-        }
+        methodComparer = MethodComparer;
```
Comment (ltrzesniewski) [inline PropertyChanged.Fody/IlGeneratedByDependencyReader.cs:19]: The `methodComparer` field is no longer needed.

File: PropertyChanged.Fody/CecilExtensions.cs (hunk 2 above)
Comment (tom-englert) [reply PropertyChanged.Fody/CecilExtensions.cs:31]: :+1:

**Comment by DaleStan** [comment]
As long as I change the base property to virtual instead of abstract, this branch looks good when I use it on the production code that spawned #635.


However, now that I've finally gotten around to porting my tests, four of them are failing. Simplified, the four failing tests look like this:
```c#
public abstract class Base : INotifyPropertyChanged
{
    public virtual string Property1 { get; set; }
    public event PropertyChangedEventHandler PropertyChanged;
}
public class Derived : Base 
{
    public override string Property1 { get; set; }
    public string Property2 => base.Property1;
}
```
The woven version of `Derived.set_Property1` calls OnPropertyChanged for both Property1 and Property2. The tests fail because they expect to see a notification only for Property1. Since `Derived.Property2` does not depend on `Derived.Property1`, I believe the tests are correct. It looks to me like `ProcessInstructionForGet` does the wrong thing when reading `Derived.get_Property2`, and I think `IsPropertyGetInstruction` needs to have different behavior for call and callvirt instructions, but I haven't figured out what the behavior for the call instruction should be.


What, if anything, do you want me to do with these tests? Should I create a PR into @tom-englert's branch?

**Comment by tom-englert** [comment]
I think the current implementation is correct regarding your example above. Since `Property1` is virtual, `this.Property1` and `base.Property1` are equivalent.
To get the behavior you are expecting, you need to use `new` instead of `override` in `Derived`.

I've already extended the tests to many of the use cases; I've also put them into one theory to get it more KISS and DRY.
If you have additional test cases that are not covered yet, just add them as cases to the `DerivedClassRaisesAllExpectedEvents` theory and create a PR against my branch.

**Comment by DaleStan** [comment]
I claim that `this.Property1` and `base.Property1` are not equivalent in my example. In `Derived`, `this.Property1` reads and writes `Derived.<Property1>k__BackingField`, while `base.Property1` reads and writes `Base.<Property1>k__BackingField`.

That said, I can't come up with a good reason to write that sort of code, so an "if it hurts, don't do that" note in the documentation seems perfectly adequate.

**Comment by tom-englert** [comment]
> "if it hurts, don't do that"

I think we could generalize this like "if you claim to be a good software developer, don't do that" :smirk: 

Obfuscation is **NOT** a design pattern.

**Pull request merged**